File: leetcode/validate-binary-search-tree.py
// https://leetcode.com/problems/validate-binary-search-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        INF = 10**10
        s = [(root, -INF, INF)]
        while s:
            node, left, right = s.pop()
            
            if node:
                if node.val <= left or node.val >= right:
                    return False
                
                s.append((node.left, left, node.val))
                s.append((node.right, node.val, right))

        
        return True

File: leetcode/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* node, long lo=LONG_MIN, long hi=LONG_MAX) {
        if (!node) return true;
        if (node->val <= lo || node->val >= hi)
            return false;
        return isValidBST(node->left, lo, node->val) && isValidBST(node->right, node->val, hi);
    }
};